File: kythe/typescript/languageserver/languageserver/src/document.ts
/*
 * Copyright 2017 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as diff from 'diff-match-patch';
import {Position, Range} from 'vscode-languageserver';

import {KytheTicketString} from './pathContext';

export interface RefResolution {
  range: Range;
  target: KytheTicketString;
  newRange?: Range;
}

// Provides a numeric value representing the size of the range. This value is
// weighted to exaggerate line differences
function size(range: Range): number {
  return (range.end.line - range.start.line) * 1000 +
      (range.end.character - range.start.character);
}

export class Document {
  // Tracks whether refs need to be regenerated because the dirty state has
  // changed
  private staleRefs = true;

  constructor(
      private refs: RefResolution[], private sourceText: string,
      private dirtyText: string = sourceText) {
    // Sort by size so we get quick resolution on lookups
    refs.sort(
        (a, b) => b === undefined ? 1 :
                                    a.range.start.line - b.range.start.line ||
                a.range.start.character - b.range.start.character);
  }

  xrefs(loc: Position): KytheTicketString|undefined {
    if (this.staleRefs) {
      this.generateDirtyRefs();
    }

    return this.refs.filter(r => rangeContains(range(r), loc))
        .sort((a, b) => size(range(a)) - size(range(b)))
        .map(r => r.target)[0];
  }

  // Takes a range from the original file in Kythe and returns that range mapped
  // into the dirty state
  dirtyRange(r: Range): Range|Error {
    if (this.staleRefs) {
      this.generateDirtyRefs();
    }

    const ref = this.refs.find(
        ref => ref.range.start.character === r.start.character &&
            ref.range.start.line === r.start.line &&
            ref.range.end.line === r.end.line &&
            ref.range.end.character === r.end.character);
    if (ref === undefined) {
      return new Error(`Unable to map range: ${r}`);
    }
    return range(ref);
  }

  updateDirtyState(text: string) {
    this.dirtyText = text;
    this.staleRefs = true;
  }

  // Diffs the current dirtyText and the sourceText, generating the shifted
  // ranges for all the refs in the file
  private generateDirtyRefs() {
    const differ = new diff.diff_match_patch();
    const diffs = differ.diff_main(this.sourceText, this.dirtyText);
    differ.diff_cleanupSemantic(diffs);
    const oldPosition: Position = {line: 0, character: 0};

    const newPosition: Position = {line: 0, character: 0};

    let refIndex = 0;

    for (const d of diffs) {
      // If we're past the ref with respect to the old file, shift refIndex
      while (refIndex < this.refs.length &&
             (this.refs[refIndex].range.start.line < oldPosition.line ||
              (this.refs[refIndex].range.start.line === oldPosition.line &&
               this.refs[refIndex].range.start.character <
                   oldPosition.character))) {
        refIndex++;
      }

      if (this.refs.length <= refIndex) {
        break;
      }

      // split the diff on lines
      const ldiff = d[1].split('\n');
      const newLine = ldiff.length > 1;
      const charOffset = ldiff[ldiff.length - 1].length;
      const diffLineLength = ldiff.length - 1;

      switch (d[0]) {
        case diff.DIFF_EQUAL:
          // The end of the range in the old file which has remained the same
          const oldEndPosition = {
            line: oldPosition.line + diffLineLength,
            character: newLine ? charOffset : oldPosition.character + charOffset
          };

          const diffRange = {start: oldPosition, end: oldEndPosition};

          // Loop through until we find a ref that starts after the diff
          for (let i = refIndex; i < this.refs.length &&
               rangeContains(diffRange, this.refs[i].range.start);
               i++) {
            // If this ref extends past the end of the diff it can't be used
            if (!rangeContains(diffRange, this.refs[i].range.end)) {
              continue;
            }

            const startLine = newPosition.line +
                (this.refs[i].range.start.line - oldPosition.line);
            const onDifferentLine = startLine !== newPosition.line;
            const refLineLength =
                this.refs[i].range.end.line - this.refs[i].range.start.line;

            // If the ref starts on a different line, newPosition.character
            // won't affect the starting character of the shifted ref
            const startChar = onDifferentLine ?
                this.refs[i].range.start.character :
                newPosition.character +
                    (this.refs[i].range.start.character -
                     oldPosition.character);

            const refCharLength =
                (this.refs[i].range.end.character -
                 this.refs[i].range.start.character);

            this.refs[i].newRange = {
              start: {line: startLine, character: startChar},
              end: {
                line: startLine + refLineLength,
                // If the ref is a single line, just add the length to the
                // startChar
                character: refLineLength > 0 ?
                    this.refs[i].range.end.character :
                    startChar + refCharLength
              }
            };
          }

        // If anything was deleted, we've made progress through the old file
        // but haven't moved our new position
        case diff.DIFF_DELETE:
          if (newLine) {
            oldPosition.line += diffLineLength;
            oldPosition.character = charOffset;
          } else {
            oldPosition.character += charOffset;
          }
          if (d[0] === diff.DIFF_DELETE) {
            break;
          }

        // If anything was inserted, we've made progress through the new file
        // but haven't moved our old position
        case diff.DIFF_INSERT:
          if (newLine) {
            newPosition.line += diffLineLength;
            newPosition.character = charOffset;
          } else {
            newPosition.character += charOffset;
          }
          break;


        default:
          break;
      }
    }

    this.staleRefs = false;
  }
}


function range(ref: RefResolution): Range {
  return ref.newRange || ref.range;
}

function rangeContains(range: Range, pos: Position) {
  // Some editors (i.e. vscode) use the character after an identifier for
  // positions so we treat ranges as inclusive on both ends
  return (range.start.line < pos.line ||
          (range.start.line === pos.line &&
           range.start.character <= pos.character)) &&
      (range.end.line > pos.line ||
       (range.end.line === pos.line && range.end.character >= pos.character));
}

File: kythe/go/languageserver/document.go
/*
 * Copyright 2017 The Kythe Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package languageserver

import (
	"sort"
	"strings"

	"github.com/sergi/go-diff/diffmatchpatch"
	"github.com/sourcegraph/go-langserver/pkg/lsp"
)

// document encapsulates all information required to map file locations to Kythe tickets
type document struct {
	refs      []*RefResolution
	oldSrc    string
	newSrc    string
	staleRefs bool
	defLocs   map[string]*lsp.Location
}

func newDocument(refs []*RefResolution, oldSrc string, newSrc string, defLocs map[string]*lsp.Location) *document {
	d := &document{
		refs:      refs,
		oldSrc:    oldSrc,
		newSrc:    newSrc,
		staleRefs: true,
		defLocs:   defLocs,
	}

	sort.Slice(d.refs, func(i, j int) bool {
		return posLess(d.refs[i].oldRange.Start, d.refs[j].oldRange.Start)
	})

	return d
}

// xrefs produces a Kythe ticket corresponding to the entity at a given
// position in the file
func (doc *document) xrefs(pos lsp.Position) *RefResolution {
	if doc.staleRefs {
		doc.generateNewRefs()
	}

	var smallestValidRef *RefResolution
	for _, r := range doc.refs {
		if r.newRange != nil &&
			rangeContains(*r.newRange, pos) &&
			(smallestValidRef == nil || smaller(*r.newRange, *smallestValidRef.newRange)) {
			smallestValidRef = r
		}
	}
	return smallestValidRef
}

// updateSource accepts new file contents to be used for diffing when next
// required This invalidates the previous diff
func (doc *document) updateSource(newSrc string) {
	doc.newSrc = newSrc
	doc.staleRefs = true
}

// rangeInNewSource takes in a range representing a ref in the oldSrc
// and returns that refs range in the newSrc if it exists
func (doc *document) rangeInNewSource(r lsp.Range) *lsp.Range {
	if doc.staleRefs {
		doc.generateNewRefs()
	}

	for _, ref := range doc.refs {
		if ref.oldRange == r {
			return ref.newRange
		}
	}

	return nil
}

// generateNewRefs generates refs by diffing the current file contents against
// the old contents
func (doc *document) generateNewRefs() {
	defer func() { doc.staleRefs = false }()

	// Short circuit if there are no references
	if len(doc.refs) == 0 {
		return
	}
	// Invalidate all previously calculated ranges
	for _, r := range doc.refs {
		r.newRange = nil
	}

	dmp := diffmatchpatch.New()
	diffs := dmp.DiffCleanupSemanticLossless(dmp.DiffMain(doc.oldSrc, doc.newSrc, true))

	// oldPos & newPos track progress through the oldSrc and newSrc respectively
	var oldPos, newPos lsp.Position
	// refIdx represents progress through the refs slice
	refIdx := 0

diffLoop:
	for _, d := range diffs {
		// We know that if our position in the old file is greater than
		// start of the ref then the current diff cannot be used to produce
		// a new range. Because refs is sorted by the start location of the
		// oldRange, we can loop forward until the current ref is past oldPos,
		// invalidating the refs along the way
		for posLess(doc.refs[refIdx].oldRange.Start, oldPos) {
			refIdx++
			// If all refs have been adjusted, nothing else needs to be done
			if len(doc.refs) <= refIdx {
				break diffLoop
			}
		}

		// Because refs are stored as lsp locations which are essentially
		// (line, character) tuples, knowing the lines contained within the
		// diff segment will be relevant
		dLines := strings.Split(d.Text, "\n")
		// If there's no newline we can't say the diff spans any lines so we
		// subtract 1
		dLineLen := len(dLines) - 1
		dNewLine := dLineLen != 0
		// dOffset determines the amount of characters the last line of the diff
		// contains
		dOffset := len(dLines[dLineLen])

		switch d.Type {
		// If text was deleted, we "move past" it in the oldSrc so we move the
		// oldPos forward
		case diffmatchpatch.DiffDelete:
			// If there was new line in the diff, then we moved forward that many lines
			// and are now at the character offset of the last line. Otherwise, we've moved
			// further through the same line
			if dNewLine {
				oldPos.Line += dLineLen
				oldPos.Character = dOffset
			} else {
				oldPos.Character += dOffset
			}

		// If text was inserted, we have "move past" it in the newSrc so we move the
		// newPos forward
		case diffmatchpatch.DiffInsert:
			if dNewLine {
				newPos.Line += dLineLen
				newPos.Character = dOffset
			} else {
				newPos.Character += dOffset
			}

		// DiffEqual is the only case where can actually map refs because we know all
		// refs contained within the equality are preserved
		case diffmatchpatch.DiffEqual:
			var oldEndChar int
			if dNewLine {
				oldEndChar = dOffset
			} else {
				oldEndChar = oldPos.Character + dOffset
			}

			// This range represents the span of the diff in the oldSrc
			diffRange := lsp.Range{
				Start: oldPos,
				End: lsp.Position{
					Line:      oldPos.Line + dLineLen,
					Character: oldEndChar,
				},
			}

			// Declare iteration variables outside because i will become the new refIdx
			// when we break or finish
			var (
				ref *RefResolution
				i   int
			)
			// Loop over the remaining unprocessed refs
			for i, ref = range doc.refs[refIdx:] {
				// When the start position is past the diffRange we know all
				// refs within the equality have been updated
				if !rangeContains(diffRange, ref.oldRange.Start) {
					break
				}

				// If the ref extends beyond the diffRange, we know the ref will
				// be invalidated
				if !rangeContains(diffRange, ref.oldRange.End) {
					continue
				}

				var (
					refStartLine  = newPos.Line + (ref.oldRange.Start.Line - oldPos.Line)
					refOnDiffLine = refStartLine != newPos.Line
					refLineLength = ref.oldRange.End.Line - ref.oldRange.Start.Line
					refCharLength = ref.oldRange.End.Character - ref.oldRange.Start.Character
					refStartChar  int
					refEndChar    int
				)

				// If the ref is on a different line than newPos, we know a newline occurs
				// between newPos and the start of the ref, meaning the start character will
				// be unchanged. Otherwise we add the offset of the ref from oldPos to newPos
				if refOnDiffLine {
					refStartChar = ref.oldRange.Start.Character
				} else {
					refStartChar = newPos.Character + ref.oldRange.Start.Character - oldPos.Character
				}

				if refLineLength > 0 {
					refEndChar = ref.oldRange.End.Character
				} else {
					refEndChar = refStartChar + refCharLength
				}

				ref.newRange = &lsp.Range{
					Start: lsp.Position{Line: refStartLine, Character: refStartChar},
					End: lsp.Position{
						Line:      refStartLine + refLineLength,
						Character: refEndChar,
					},
				}
			}

			// We know there's no reason to go back to any refs
			// within the diffRange
			refIdx += i

			// On an equality we have to move both oldPos and newPos forward
			if dNewLine {
				oldPos.Line += dLineLen
				oldPos.Character = dOffset
				newPos.Line += dLineLen
				newPos.Character = dOffset
			} else {
				oldPos.Character += dOffset
				newPos.Character += dOffset
			}
		}
	}
}

// RefResolution represents the mapping from a location in a document to a
// Kythe ticket
type RefResolution struct {
	ticket   string
	def      string     // the target definition anchor ticket
	markup   string     // a rendering of marked source
	comment  string     // if available, a comment
	lang     string     // a language label
	oldRange lsp.Range  // the range indexed
	newRange *lsp.Range // the range after patching (if viable)
}

func posLess(a, b lsp.Position) bool {
	if a.Line == b.Line {
		return a.Character < b.Character
	}
	return a.Line < b.Line
}

func posLeq(a, b lsp.Position) bool {
	if a.Line == b.Line {
		return a.Character <= b.Character
	}
	return a.Line < b.Line
}

// rangeContains checks whether a Range contains a Position. This
// check is inclusive because the position represents a cursor location.
// Consider an identifier "hi" at offset 0. Cursor positions 0-2 inclusive
// should match it.
func rangeContains(r lsp.Range, pos lsp.Position) bool {
	return posLeq(r.Start, pos) && posLeq(pos, r.End)
}

func smaller(r1, r2 lsp.Range) bool {
	r1lines := r1.End.Line - r1.Start.Line
	r2lines := r2.End.Line - r2.Start.Line

	if r1lines == r2lines {
		r1chars := r1.End.Character - r1.Start.Character
		r2chars := r2.End.Character - r2.Start.Character
		return r1chars < r2chars
	}
	return r1lines < r2lines
}
